torch/distributed/fsdp/flat_param.py

```diff
@@ -311,22 +343,33 @@ def _init_flat_param(
                         )
                     )
                 else:
-                    if type(param) is FlatParameter
+                    if type(param) is FlatParameter:
                         raise ValueError("`FlatParameter` does not support nesting")
                     if dtype is not None and param.dtype != dtype:
                         raise ValueError(
                             "`FlatParameter` requires uniform dtype but got "
                             f"{dtype} and {param.dtype}"
                         )
-                    if dtype is None and not param.is_floating_point():
-                        raise ValueError("Integer parameters are unsupported")
+                    if dtype is None:
+                        try:
```

Comment:
This is not good. `ShardedTensor` does not support `is_floating_point()`, so we get an error in `__torch_function__()`:
https://github.com/pytorch/pytorch/blob/270e5e519d98868af0166f3a179b286682cfb267/torch/distributed/_shard/sharded_tensor/api.py#L1085-L1087